File: backend/game/TicTacToe.py
class TicTacToe:
    def __init__(self):
        self.board = [' '] * 9
        self.current_player = 'X'

    def available_actions(self):
        return [i for i, v in enumerate(self.board) if v == ' ']
    
    def take_action(self, action):
        self.board[action] = self.current_player

        if self.check_win():
            return 1 if self.current_player == 'X' else -1
        elif ' ' not in self.board:
            return 0
        else:
            self.current_player = 'O' if self.current_player == 'X' else 'X'
            return None
        
    def check_win(self):
        combinations = [
            [0, 1, 2], [3, 4, 5], [6, 7, 8], # rows
            [0, 3, 6], [1, 4, 7], [2, 5, 8], # columns
            [0, 4, 8], [2, 4, 6]             # diagonals
        ]

        for combo in combinations:
            if self.board[combo[0]] == self.board[combo[1]] == self.board[combo[2]] != ' ':
                return True
        
        return False

    def reset(self):
        self.board = [' '] * 9
        self.current_player = 'X'

File: src/components/TicTacToe.js
import React, { useState } from "react";

function TicTacToe() {
    const [board, setBoard] = useState(Array(9).fill(null));
    const [currentPlayer, setCurrentPlayer] = useState("X");
    const [winner, setWinner] = useState(null);

    const handleCellClick = (index) => {
        if (winner || board[index]){
            return;
        }

        const newBoard = [...board];
        newBoard[index] = currentPlayer;

        setBoard(newBoard);

        const isDraw = checkForDraw(newBoard);
        const newWinner = checkForWinner(newBoard);
        if(isDraw){
            setWinner('draw');
            // sleep for 1 sec and reset
            setTimeout(() => resetGame(), 1500);
            return;
        }

        if(newWinner){
            setWinner(newWinner);
        }else{
            setCurrentPlayer(currentPlayer === "X" ? "O" : "X");
        }
    };

    const checkForWinner = (board) => {
        // list all winnable combinations
        const COMBINATIONS = [
            [0,1,2],[3,4,5],[6,7,8], //rows
            [0,3,6],[1,4,7],[2,5,8], //columns
            [0,4,8],[2,4,6]
        ];

        for(let combo of COMBINATIONS){
            const [a,b,c] = combo;
            if(board[a] && board[a] === board[b] && board[a] === board[c]){
                console.log(board[a]);
                return board[a];
            }
        }

        return null;
    };

    const checkForDraw = (board) => {
        for(let x of board){
            console.log(x);
            if(x == null){
                return false;
            }
        }

        if(!checkForWinner(board)){
            return true;
        }

    };

    const resetGame = () => {
        setBoard(Array(9).fill(null));
        setWinner(null);
        setCurrentPlayer('X');
    };


    return (
        <div className="App">
            <header className="App-header">
                <h1>Tic Tac Toe</h1>
                    <div className="row">
                        <div className="board">
                            {
                            board.map((cell, index) => (
                                <div key={index} className="cell" onClick={() => handleCellClick(index)}>{cell}</div>
                            ))
                            }
                        </div>
                        <div className="column">
                            <div className="text">
                                {winner ? `Winner: ${winner}` : `Next Player: ${currentPlayer}`}
                            </div>
                            <button className="reset_button" onClick={resetGame}>Reset Game</button>
                        </div>
                    </div>
            </header>
        </div>
    );
}

export default TicTacToe;